### ISSUE TITLE
Bug: [dot-notation] False positive with optional chaining and allowPrivateClassPropertyAccess: true

### ISSUE THREAD

**Issue by y-hsgw** [typescript-eslint/typescript-eslint#8813]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.4.3&fileType=.ts&code=MYGwhgzhAEAa0G8BQ1oAcBOBLAbmALgKbrY4D6mA9mtALzQCMATAMwDcSAvkkiIftAAeALjjQAPtACuAOwAmhAGZYZhOR2CUZESnwB0ISgHMAFIID8egNoAiTLgoZqNgLoBKJEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQCYHsAXAWgDsiBDQgS31PSnwDMnIAacbSAAUIE8ADigDG0agJIp41UoQD0BEuUJVa9DAG1O2KImjR80dtuxYdXCvHj4A7gAUxANyqIAwvArJkD-EOj8AQWFhFFQMQjhEEwgAXxMAXU44mKA&tsconfig=&tokens=false

### Repro Code

```TypeScript
class X {
  private priv_prop = 123;
}

let x: X | undefined;
// ["priv_prop"] is better written in dot notation.
console.log(x?.["priv_prop"])
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "dot-notation": "off",
    "@typescript-eslint/dot-notation": [
      "error",
      {
        "allowPrivateClassPropertyAccess": true
      }
    ]
  }
}
```

### Expected Result

Expect no ESLint errors.

### Actual Result

This ESLint error.
`["priv_prop"] is better written in dot notation.`

### Additional Info

Similar issues：https://github.com/typescript-eslint/typescript-eslint/issues/3510
